lib/gen_debouncer.ex

```diff
@@ -0,0 +1,169 @@
+defmodule GenDebouncer do
+  use GenServer
+
+  defmodule State do
+    @moduledoc false
+    @enforce_keys [:worker, :interval]
```

Comment:
NIT: I think the most commonly used term for the defined time to wait before executing work i've seen is `timeout` rather than `interval`

Reply:
If we use send_interval, we will have a constant interval, no matter of the incoming events that should move the execution further down in time.

So, if new work gets added it won't execute because the wait time is not enough (that's ok), but the time to the next tick won't change, so when the conditions are finally met, we could have been waiting for more than the actual time we are trying to get.

I think the debouncer logic is clasically modeled as a timout that gets canceled/refreshed, and when if/it triggers is because the condition was met :thinking_face: [](edited-with-Axolo)

Reply:
I'm wondering if we should do something like that instead, maybe a timer per each key in the state? [](edited-with-Axolo)

Reply:
With our use case it does not matter. We have two options currently:
• serial execution, use the GenDebouncer thread to execute
• async execution, still would need to wait for result
• fire and forget execution
• let use determine that, just execute the code inline and document that it should not take long time i.e. only send messages, only spawn tasks etc. [](edited-with-Axolo)